staging/src/k8s.io/component-base/config/serializer/encode.go

```diff
@@ -0,0 +1,260 @@
+	return cs.codecs.CodecForVersions(nil, cs.strictDeserializer, nil, runtime.InternalGroupVersioner)
+}
+
+// DecodeInto decodes bytes into a pointer of the desired type
+func (cs *configSerializer) DecodeInto(data []byte, into runtime.Object) error {
+	out, gvk, err := cs.universalDecoder().Decode(data, nil, into)
+	// Return a structured error if the group was registered with the scheme but the version was too old
+	if gvk != nil && err != nil {
+		if cs.scheme.IsGroupRegistered(gvk.Group) && !cs.scheme.IsVersionRegistered(gvk.GroupVersion()) {
+			msg := fmt.Sprintf("unrecognized version %s in known group %s", gvk.Version, gvk.Group)
+			return NewUnrecognizedAPIVersionError(msg, *gvk, data)
+		}
+	}
+	if err != nil {
+		return err
+	}
+	if out != into {
+		return fmt.Errorf("unable to decode %s into %v", gvk, reflect.TypeOf(into))
+	}
+	return nil
+}
+
+// DecodeManyInto decodes bytes into a pointer of the desired type
+func (cs *configSerializer) DecodeManyInto(src EncodedConfigs, dest DecodedConfigs) error {
+	for gvk, file := range src {
+		into, ok := dest[gvk]
+		if !ok {
+			return fmt.Errorf("src and dest aren't equal")
+		}
+		if err := cs.DecodeInto(file, into); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Encode encodes the object into a byte slice for the specific format and version
+func (cs *configSerializer) Encode(format EncodingFormat, gv schema.GroupVersion, obj runtime.Object) ([]byte, error) {
+	encoder, ok := cs.mediaTypeMap[format]
+	if !ok {
+		return []byte{}, fmt.Errorf("encoding format not supported: %s", format)
+	}
+	// versionSpecificEncoder writes out YAML bytes for exactly this group version
+	versionSpecificEncoder := cs.codecs.EncoderForVersion(encoder, gv)
+	// Encode the object to YAML for the given version
+	return runtime.Encode(versionSpecificEncoder, obj)
+}
+
+// EncodeMany encodes multiple inputs in a gvk-object map into a byte slice where the objects are encoded as multiple YAML documents
+func (cs *configSerializer) EncodeMany(format EncodingFormat, objs DecodedConfigs) ([]byte, error) {
+	docs := EncodedConfigs{}
+	for gvk, obj := range objs {
+		b, err := cs.Encode(format, gvk.GroupVersion(), obj)
+		if err != nil {
+			return nil, err
+		}
+		if err := docs.Add(gvk, b); err != nil {
+			return nil, err
+		}
+	}
+	return docs.GetAllDocuments(), nil
+}
+
+// EncodingFormat is a high-level, typed constant configuring how an object should be encoded
+type EncodingFormat string
+
+const (
+	// YAML encodes a config file in the YAML format
+	YAML EncodingFormat = "YAML"
+	// JSON encodes a config file in compact JSON format
+	JSON EncodingFormat = "JSON"
+	// JSONPretty encodes a config file in pretty JSON format
+	JSONPretty EncodingFormat = "JSONPretty"
+)
+
+// DecodedConfigs represents a collection of decoded configuration files.
+// The objects are indexed by their GroupVersionKind.
+type DecodedConfigs map[schema.GroupVersionKind]runtime.Object
```

Comment:
I'm not sure a 1:1 type to instance relationship is so fundamental to configuration we should make it part of the serializer. can we wait to add this until it is clear this is actually fundamental?